ESP32-C3 USB serial driver

martinjaeger opened the pull request:
This is a first attempt to get support for the built-in USB serial of the ESP32-C3. It is not a full-featured USB interface, that's why I put the drivers under `serial` instead of implementing them on top of the USB stack as CDC-ACM.

This is a single-instance driver, so we don't need the `const struct device *dev` in most of the functions.

Currently the driver does not compile because the header `hal/usb_serial_jtag_ll.h` is not existing in `hal_espressif`, but only in the original `esp-idf`.

@sylvioalves Could you have a look if the general approach of the driver looks fine or if there is a more suitable way to support this special serial interface? And which is the correct way to get the headers from ESP-IDF into `hal_espressif`?

martinjaeger committed:
drivers: serial: esp32_usb: Add USB serial / JTAG interface

This peripheral combines a hardware-based USB CDC ACM serial interface
and a JTAG interface.

It is present in the ESP32-C3.

Signed-off-by: Martin Jäger <martin@libre.solar>

sylvioalves commented:
@martinjaeger, sure, I'll take a look. `hal_espressif` will be update to 4.4.1 soon, which will add missing dependencies.

martinjaeger commented:
@sylvioalves Sounds good. I got it working with some dirty hacks (copying some files from ESP-IDF to locations where they don't belong). Just let me know when the `hal_espressif` update has been finished and I'll update the PR so that you can review a working version.

lgl88911 commented:
@sylvioalves After hal_espressif update to 4.4.1, Whether to support west flash via USB serial of esp32c3?

martinjaeger commented:
@sylvioalves @glaubermaroto can you give any update on this?

If you don't have the time to update `hal_espressif` I'm fine with that and I'll come up with a workaround. But I'd really like to make some progress here.

@lgl88911 Yes, flashing via USB serial should work after this has been finished. I've successfully used it in my fork already.

martinjaeger commented:
@lgl88911 FYI, here is an out-of-tree driver implementation which works very well (both for flashing and serial input/output): https://github.com/LibreSolar/bms-firmware/commit/1f1a53b33451bb4206a10fe00faeb92c69514605

lgl88911 commented:
> @lgl88911 FYI, here is an out-of-tree driver implementation which works very well (both for flashing and serial input/output): [LibreSolar/bms-firmware@1f1a53b](https://github.com/LibreSolar/bms-firmware/commit/1f1a53b33451bb4206a10fe00faeb92c69514605)

That's great, before the main branch is modified, it is painless to use USB serial. Also ask me if I need to upgrade hal_espressif and esp-idf to use it? In my understanding, esp32c3 needs specific esp-idf support to use USB serial

martinjaeger commented:
> @martinjaeger, can you remove the "WIP:" from commit message and add missing sign-off id?

Sure. Was the HAL updated already without me noticing so that it should work now? If yes, I can clean up this PR and make it ready for review.

sylvioalves commented:
Not yet, I'll bring those USB files before submitting the whole HAL update.

> 

PTAL https://github.com/zephyrproject-rtos/hal_espressif/pull/118

martinjaeger commented:
> Not yet, I'll bring those USB files before submitting the whole HAL update.

Excellent, thanks. Please note also my other PR for the UART1 pinctrl. I'd like to backport that fix to v3.1-branch, so it would be nice if that could get merged quickly as well.

martinjaeger committed:
dts: riscv: esp32c3: add usb_serial devicetree node

Devicetree configuration for ESP32-C3 USB serial / JTAG interface.

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger committed:
boards: riscv: esp32c3_devkit_m: enable USB serial

Enable the USB serial / JTAG interface so that it is compiled during
build tests.

Signed-off-by: Martin Jäger <martin@libre.solar>

sylvioalves commented:
@martinjaeger Is anything else needed to complete this?

martinjaeger commented:
https://github.com/zephyrproject-rtos/zephyr/pull/46668 has to be merged.

carlescufi reviewed (dismissed):
LGTM

carlescufi commented:
@martinjaeger please rebase

martinjaeger commented:
> @martinjaeger please rebase

Sorry, currently on vacation without laptop. If it's possible for someone else to rebase the PR, I'd appreciate that. Otherwise will do as soon as I'm back.

fabiobaltieri commented:
> > @martinjaeger please rebase
> 
> Sorry, currently on vacation without laptop. If it's possible for someone else to rebase the PR, I'd appreciate that. Otherwise will do as soon as I'm back.

Hey I rebased for you, easy enough, trivial conflict in [esp32c3.dtsi](https://github.com/zephyrproject-rtos/zephyr/pull/44045/files#diff-4ef0770e2844cf23c1eb6976bfb64e1c127c4f664fcaa9b5babc4e4919866d2f)

gmarull commented on drivers/serial/serial_esp32_usb.c:
```diff
@@ -0,0 +1,290 @@
+
+#define ESP32_UART_INIT(idx)                                                                       \
+	static const DRAM_ATTR struct serial_esp32_usb_config serial_esp32_usb_cfg_port_##idx = {  \
+		.uart_num = DT_INST_PROP(idx, peripheral),                                         \
```
I can't see this being used

gmarull commented on drivers/serial/serial_esp32_usb.c:
```diff
@@ -0,0 +1,290 @@
+#include <zephyr/sys/util.h>
+#include <esp_attr.h>
+
+#define ISR_HANDLER isr_handler_t
```
nit: why is this needed? can't we just do `(isr_handler_t)`?

gmarull commented on drivers/serial/serial_esp32_usb.c:
```diff
@@ -0,0 +1,290 @@
+
+#define DT_DRV_COMPAT espressif_esp32_usb_serial
+
+#include "stubs.h"
```
doesn't seem to be a local header, please use `<>`

gmarull commented:
too late :-( please check my comments anyway and send a follow-up PR where applicable.

martinjaeger commented:
Thanks @gmarull for your thorough review. There were indeed some left-overs from porting the "normal" UART driver to the special USB serial peripheral.

See follow-up PR with some additional clean-up: #47750
